### PR TITLE
UndoManager: Make undo() and redo() asynchronous

### PR DESCRIPTION
**Pull request by tintou**

### CODE REVIEW

File: libcore/marlin-undostack-manager.c [hunk 1]
```diff
@@ -370,18 +370,21 @@ marlin_undo_manager_is_undo_redo (MarlinUndoManager *manager)
  * Redoes the last file operation
 ** ****************************************************************/
 void
-marlin_undo_manager_redo (MarlinUndoManager *self,
-                          GtkWidget *parent_view,
-                          MarlinUndoFinishCallback cb,
-                          gpointer callback_data)
+marlin_undo_manager_redo (MarlinUndoManager   *self,
```
Comment (jeremypw) [inline libcore/marlin-undostack-manager.c]: Maybe rename `marlin_undo_manager_redo_async` ?

File: libcore/marlin-undostack-manager.c [hunk 2]
```diff
@@ -481,25 +484,39 @@ marlin_undo_manager_redo (MarlinUndoManager *self,
...
 
 /** ****************************************************************
  * Undoes the last file operation
 ** ****************************************************************/
 void
-marlin_undo_manager_undo (MarlinUndoManager *self,
-                          GtkWidget *parent_view,
-                          MarlinUndoFinishCallback cb,
-                          gpointer done_callback_data)
+marlin_undo_manager_undo (MarlinUndoManager   *self,
```
Comment (jeremypw) [inline libcore/marlin-undostack-manager.c]: Rename `..._async`?

**Comment by jeremypw** [comment]
Is there anything new we need to inhibit in the main context while undo/redo is ongoing asynchronously?

**Comment by jeremypw** [comment]
Are there circumstances where the undo/redo needs to be cancelled?  Is it a good idea to cancel an undo half-way through?  What does it mean?  Should those parts of the undo completed be reversed?

**Comment by jeremypw** [comment]
Where the undo operation is itself async (e.g. UNDO_MOVE) the task completion callback triggers *before* the operation actually completes.  Is this intended?  This causes the "after_undo_redo" Window function to be called while the file operations are ongoing which does not seem right.  Although, checking the master code I think the same issue occurs there.  I think the MarlinUndoFinishCallback fuinction should be called in the `undo_redo_transfer_callback` function when appropriate.  I am not sure why some file operations are async and other user sync GLib functions (which would not show and progress window).  Its a long time since I looked at this part of the code :-/

**Comment by tintou** [comment]
> Where the undo operation is itself async (e.g. UNDO_MOVE) the task completion callback triggers _before_ the operation actually completes. Is this intended? This causes the "after_undo_redo" Window function to be called while the file operations are ongoing which does not seem right. Although, checking the master code I think the same issue occurs there. I think the MarlinUndoFinishCallback fuinction should be called in the `undo_redo_transfer_callback` function when appropriate. I am not sure why some file operations are async and other user sync GLib functions (which would not show and progress window). Its a long time since I looked at this part of the code :-/

I tried to keep the same behavior as in master, I agree that the completion is asynchronous but that can be reworked later :)

**Comment by tintou** [comment]
> Are there circumstances where the undo/redo needs to be cancelled? Is it a good idea to cancel an undo half-way through? What does it mean? Should those parts of the undo completed be reversed?

In some cases it might make sense, if you copy a folder over a high-latency network and undo the operation you also might want to cancel the operation I assume. We might want to give a NULL value here for the time being, but having it supported in some places might be a good thing regarding the future

File: libcore/marlin-undostack-manager.c (hunk 1 above)
Comment (tintou) [reply libcore/marlin-undostack-manager.c]: It is not mandatory to call it *_async as there is a *_finish anyway so whatever you perfer

Comment (jeremypw) [reply libcore/marlin-undostack-manager.c]: I guess it only matters where there are two similar functions, one sync and one async like in Gio so just leave it.

**Pull request merged**